nexus-repository-r/src/main/java/org/sonatype/nexus/repository/r/internal/RComponentDirector.java

```diff
@@ -0,0 +1,110 @@
+
+import java.util.List;
+import java.util.Optional;
+
+import javax.inject.Inject;
+import javax.inject.Named;
+import javax.inject.Singleton;
+
+import org.sonatype.goodies.common.ComponentSupport;
+import org.sonatype.nexus.common.entity.EntityId;
+import org.sonatype.nexus.common.entity.EntityMetadata;
+import org.sonatype.nexus.common.event.EventManager;
+import org.sonatype.nexus.repository.Repository;
+import org.sonatype.nexus.repository.manager.RepositoryManager;
+import org.sonatype.nexus.repository.storage.Asset;
+import org.sonatype.nexus.repository.storage.AssetUpdatedEvent;
+import org.sonatype.nexus.repository.storage.Bucket;
+import org.sonatype.nexus.repository.storage.BucketStore;
+import org.sonatype.nexus.repository.storage.Component;
+import org.sonatype.nexus.repository.storage.ComponentDirector;
+
+import static com.google.common.base.Preconditions.checkNotNull;
+
+/**
+ * @since 1.next
+ */
+@Named(RFormat.NAME)
+@Singleton
+public class RComponentDirector
+    extends ComponentSupport
+    implements ComponentDirector
+{
+  private final BucketStore bucketStore;
+
+  private final RepositoryManager repositoryManager;
+
+  private final EventManager eventManager;
+
+  private String sourceRepositoryName;
+
+  @Inject
+  public RComponentDirector(final BucketStore bucketStore,
+                            final RepositoryManager repositoryManager,
+                            final EventManager eventManager)
+  {
+    this.bucketStore = checkNotNull(bucketStore);
+    this.repositoryManager = checkNotNull(repositoryManager);
+    this.eventManager = checkNotNull(eventManager);
+  }
+
+  @Override
+  public boolean allowMoveTo(final Repository destination) {
+    return true;
+  }
+
+  @Override
+  public boolean allowMoveTo(final Component component, final Repository destination) {
+    return repositoryFor(component).isPresent();
+  }
+
+  @Override
+  public boolean allowMoveFrom(final Repository source) {
+    return true;
+  }
+
+  private Optional<Repository> repositoryFor(final Component component) {
+    return Optional.of(component)
+        .map(Component::bucketId)
+        .map(bucketStore::getById)
+        .map(Bucket::getRepositoryName)
+        .map(repositoryManager::get);
+  }
+
+  @Override
+  public Component beforeMove(final Component component,
+                              final List<Asset> assets,
+                              final Repository source,
+                              final Repository destination)
+  {
+    sourceRepositoryName = source.getName();
+    return component;
+  }
+
+  @Override
+  public Component afterMove(final Component component, final Repository destination) {
+    EntityMetadata entityMetadata = component.getEntityMetadata();
+    assert entityMetadata != null;
+    EntityId componentId = entityMetadata.getId();
+
+    if (sourceRepositoryName != null) {
+      eventManager.post(new AssetUpdatedEvent(entityMetadata, sourceRepositoryName, componentId));
```

Comment:
If asset is created/deleted event should be posted automatically. Does it really necessary to manually post it here?

Reply:
By staging we do not change an asset, we change components. To fire reindex metadata which are looking on assets state I need manually call this event.